pandas/core/internals.py

```diff
@@ -2437,7 +2507,8 @@ class DatetimeBlock(DatetimeLikeBlockMixin, Block):
     _can_hold_na = True
 
     def __init__(self, values, placement, ndim=None):
-        if values.dtype != _NS_DTYPE:
+        if values.dtype != _NS_DTYPE and values.dtype.base != _NS_DTYPE:
```

Comment:
huh?

Reply:
Does the comment immediately below explain it?

Reply:
If your question was about why the change was needed, that's because `NonConsolidatableMixin` now calls `super().__init__`, so `DatetimeTZBlock.__init__` ends up here.

Reply:
my qestion is why are you accesing base, it should not be needed (so why the change)

Reply:
(without the change)

```python
In [2]: pd.Series(pd.date_range("2017", periods=4, tz="US/Central"))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-bafaac91d0ed> in <module>()
----> 1 pd.Series(pd.date_range("2017", periods=4, tz="US/Central"))

~/sandbox/pandas-ip/pandas/pandas/core/series.py in __init__(self, data, index, dtype, name, copy, fastpath)
    239                                        raise_cast_failure=True)
    240
--> 241                 data = SingleBlockManager(data, index, fastpath=True)
    242
    243         generic.NDFrame.__init__(self, data, fastpath=True)

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in __init__(self, block, axis, do_integrity_check, fastpath)
   4467
   4468         if not isinstance(block, Block):
-> 4469             block = make_block(block, placement=slice(0, len(axis)), ndim=1)
   4470
   4471         self.blocks = [block]

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in make_block(values, placement, klass, ndim, dtype, fastpath)
   3006                      placement=placement, dtype=dtype)
   3007
-> 3008     return klass(values, ndim=ndim, placement=placement)
   3009
   3010 # TODO: flexible with index=None and/or items=None

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in __init__(self, values, placement, ndim, dtype)
   2662
   2663         super(DatetimeTZBlock, self).__init__(values, placement=placement,
-> 2664                                               ndim=ndim)
   2665
   2666     def copy(self, deep=True, mgr=None):

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in __init__(self, values, placement, ndim)
   1700                 ndim = 2
   1701         super(NonConsolidatableMixIn, self).__init__(values, placement,
-> 1702                                                      ndim=ndim)
   1703
   1704     @property

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in __init__(self, values, placement, ndim)
   2514         if values.dtype != _NS_DTYPE:
   2515             # not datetime64 or datetime64tz
-> 2516             values = conversion.ensure_datetime64ns(values)
   2517
   2518         super(DatetimeBlock, self).__init__(values,

TypeError: Argument 'arr' has incorrect type (expected numpy.ndarray, got DatetimeIndex)
```

Reply:
something is wrong if you have to change this.

Reply:
As Tom said before, this is because DatetimeTZBlock also calls this init method, and they have a `DatetimeIndex` as holder. So this check needs to let a DatetimeIndex through (and not only a datetime64 dtyped array).  
Tom, is that a correct interpretation? In any case a long comment to clarify could be useful.

Reply:
How about a method on `DatetimeBlock` and `DatetimeTZBlock` that checks this?

For `DatetimeBlock`:

```
def maybe_ensure_datetime64ns(self, values):
    if values.dtype != _NS_DTYPE:
```

And in `DatetimeTZBlock`:

```
def maybe_ensure_datetime64ns(self, values):
    if values.dtype.base != _NS_DTYPE:
```

That's more indirection, but more correct?

Reply:
@jorisvandenbossche the `__init__` change is the root issue, yes.

I think defining a method on each block, like `_maybe_coerce_values` or something, may be a useful refactoring anyway. Lot's of subclasses are essentially doing this in the `__init__` anyway, so we could remove all those and just call `_maybe_coerce_values` in the parent.

Will do for Datetime now, and we can expand if we like it.

Reply:
In any case, it would be more correct. As with the change above you allow datetime tz values also for a DatetimeBlock, which should not be accepted. So I think that would be a good idea to have the check in a method that can be overridden by DatetimeTZBlock

Reply:
Sorry, confused myself again, it *is* doing coercion. We have I was looking at `datetimes._ensure_datetime64`, which does validation, not `conversion.ensure_datetime64ns`, which does coercion.

Reply:
The `dtype` argument to `DatetimeTZBlock`  makes this difficult, since no other block accepts it. If we can remove that and do the conversion at the caller, it'd be a nice cleanup. Going to try to work around it for now though and do in a followup.